src/sys/unix.rs

```diff
@@ -2012,20 +2012,44 @@ impl crate::Socket {
     }
 }
 
+#[cfg_attr(docsrs, doc(cfg(unix)))]
+impl AsFd for crate::Socket {
+    fn as_fd(&self) -> BorrowedFd<'_> {
+        // SAFETY: lifetime is bound by self.
+        unsafe { BorrowedFd::borrow_raw(self.as_raw()) }
+    }
+}
+
 #[cfg_attr(docsrs, doc(cfg(unix)))]
 impl AsRawFd for crate::Socket {
     fn as_raw_fd(&self) -> c_int {
         self.as_raw()
     }
 }
 
+#[cfg_attr(docsrs, doc(cfg(unix)))]
+impl From<crate::Socket> for OwnedFd {
+    fn from(sock: crate::Socket) -> OwnedFd {
+        // SAFETY: sock.into_raw() always returns a valid fd.
+        unsafe { OwnedFd::from_raw_fd(sock.into_raw()) }
```

Comment:
Currently your `Socket::from_raw` constructor does not guarantee this as it is still not marked unsafe.

Reply:
Could we use `from_raw_fd` here instead, since that one has an unsafe precondition that the file descriptor is valid?

Reply:
Could this be ameliorated by adding an unsafe bound on `from_raw`?

Reply:
> Currently your `Socket::from_raw` constructor does not guarantee this as it is still not marked unsafe.

As per the comment on the function: https://github.com/rust-lang/socket2/blob/057f6c2a02c9fe6cc56f705a47405d36090743cc/src/socket.rs#L80-L85

It should be marked as unsafe, but then it can be passed (easily) as a closure any more. So it should be an unsafe function, it's only not because it's an internal function.

But to be clear `crate::Socket` always holds a valid fd/socket, with the assumption that the OS will not return an invalid fd/socket (outside of errors of course). I also place to switch to [`std::os::unix::io::OwnedFd`](https://doc.rust-lang.org/std/os/unix/io/struct.OwnedFd.html) to take advantage of the layout optimisation.

Reply:
Oh, I didn't actually notice that it was an internal function.